app/models/works_generator.rb

```diff
@@ -0,0 +1,181 @@
+                             location: "#{FFaker::AddressUS.city}, #{FFaker::AddressUS.state}, United States",
+                             start_on: Date.today - 6.months,
+                             end_on: Date.tomorrow - 6.months,
+                             delivery_type: 'Invitation',
+                             scope: 'National'
+
+    FactoryBot.create :user_performance,
+                      activity_insight_id: perf.activity_insight_id,
+                      user: user,
+                      performance: perf
+  end
+
+  private
+
+    def create_duplicate_publication(publication)
+      duplicate = publication.dup
+      duplicate.update({ visible: false, duplicate_publication_group_id: publication.duplicate_publication_group_id })
+      duplicate.save
+      duplicate
+    end
+
+    def create_contributor_name(publication)
+      FactoryBot.create(:contributor_name,
+                        first_name: FFaker::Name.first_name,
+                        last_name: FFaker::Name.last_name,
+                        publication: publication)
+    end
+
+    def create_non_duplicate_group_membership(publication, non_duplicate_group_id)
+      FactoryBot.create :non_duplicate_publication_group_membership,
+                        publication: publication,
+                        non_duplicate_publication_group_id: non_duplicate_group_id
+    end
+
+    def create_activity_insight_publication_import(publication)
+      FactoryBot.create :publication_import,
+                        source: 'Activity Insight',
+                        source_identifier: rand(10000000..99999999),
+                        publication: publication
+    end
+
+    def create_pure_publication_import(publication)
+      FactoryBot.create :publication_import,
+                        source: 'Pure',
+                        source_identifier: FFaker::Lorem.characters(50),
+                        publication: publication
+    end
+
+    def create_authorship(publication)
+      FactoryBot.create :authorship,
+                        user: user,
+                        publication: publication
+    end
+
+    def rand_non_journal_article_type
+      Publication.publication_types.map { |type| type if type.exclude?('Journal Article') }.compact.sample
+    end
+
+    def rand_journal_article_type
+      Publication.publication_types.map { |type| type if /Journal Article/.match?(type) }.compact.sample
+    end
+
+    def pub_attrs
+      {
+        title: FFaker::Lorem.sentence,
+        secondary_title: FFaker::Lorem.sentence,
+        journal: rand_journal,
+        abstract: FFaker::Lorem.paragraph,
+        isbn: FFaker::Book.isbn,
+        volume: rand(50),
+        edition: rand(50),
+        doi: "https://doi.org/10.#{rand(1000..9999)}/#{FFaker::Internet.ip_v4_address}",
+        published_on: (Date.today - 6.months)
+      }
+    end
+
+    def rand_journal
+      Journal.count.zero? ? FactoryBot.create(:journal) : Journal.find(rand(Journal.count))
```

Comment:
There's a chance `Journal.find(rand(Journal.count))` could raise an exception, if a journal is ever deleted then it will leave a "hole" behind in the ids, which will raise a `RecordNotFound` exception if randomly hit. Pretty unlikely but possible.

Another way to do this same without falling into a missing id hole is `Journal.order('RANDOM()').first`